runtime/drive_handler.go

```diff
@@ -96,32 +96,80 @@ type StubDriveHandler struct {
 // Reserve pops a unused stub drive and returns a MountableStubDrive that can be
 // mounted with the provided options as the patched drive information.
 func (h *StubDriveHandler) Reserve(
+	requestCtx context.Context,
 	id string,
 	hostPath string,
 	vmPath string,
 	filesystemType string,
 	options []string,
-) (MountableStubDrive, error) {
+	driveMounter drivemount.DriveMounterService,
+	machine firecracker.MachineIface,
+) error {
 	h.mu.Lock()
 	defer h.mu.Unlock()
 	if len(h.freeDrives) == 0 {
-		return nil, ErrDrivesExhausted
+		return ErrDrivesExhausted
+	}
+	if _, ok := h.usedDrives[id]; ok {
+		// This case means that drive wasn't released or removed properly
+		return fmt.Errorf("drive with ID %s already in use, a previous attempt to remove it may have failed", id)
 	}
 
 	freeDrive := h.freeDrives[0]
 	options, err := setReadWriteOptions(options, freeDrive.driveMount.IsWritable)
 	if err != nil {
-		return nil, err
+		return err
 	}
 
-	h.freeDrives = h.freeDrives[1:]
-	h.usedDrives[id] = freeDrive
-	return freeDrive.withMountConfig(
+	stubDrive := freeDrive.withMountConfig(
 		hostPath,
 		vmPath,
 		filesystemType,
 		options,
-	), nil
+	)
+	freeDrive = &stubDrive
+
+	err = stubDrive.PatchAndMount(requestCtx, machine, driveMounter)
+	if err != nil {
+		err = errors.Wrapf(err, "failed to mount drive inside vm")
+		return err
+	}
+
+	h.freeDrives = h.freeDrives[1:]
+	h.usedDrives[id] = freeDrive
+	return nil
+}
+
+// Release unmounts stub drive of just deleted container
+// and pushes just released drive to freeDrives
+func (h *StubDriveHandler) Release(
+	requestCtx context.Context,
+	id string,
+	driveMounter drivemount.DriveMounterService,
+	machine firecracker.MachineIface,
+) error {
+	h.mu.Lock()
+	defer h.mu.Unlock()
+	drive, ok := h.usedDrives[id]
+	if !ok {
+		return errors.Errorf("container %s drive wasn't found", id)
+	}
+
+	_, err := driveMounter.UnmountDrive(requestCtx, &drivemount.UnmountDriveRequest{
+		DriveID: drive.driveID,
+	})
+	if err != nil {
+		return errors.Wrap(err, "failed to unmount drive")
+	}
+
+	err = machine.UpdateGuestDrive(requestCtx, drive.driveID, drive.stubPath)
+	if err != nil {
+		return errors.Wrap(err, "failed to patch drive")
+	}
+
```

Comment:
I just took a look at the errors in buildkite in depth, they are more complicated than I was originally thinking, but there is an easy fix that will work in the short-term: change `drive.stubPath` to `filepath.Base(drive.stubPath)`.

The problem here is that there's some pre-existing implicit dependencies between the jailer and this drive handler. The error that was occurring in buildkite was due to the fact that here we specify a full path on the host (i.e. `/var/lib/fccd/<vmID>/rootfs/stub0`), but a jailed firecracker binary lives inside a rootfs where paths for the same file are different (i.e. `/stub0`).

The reason just using `filepath.Base(drive.stubPath)` fixes the problem is because that will set it to the correct path for both the no-op jailer and the [runc jailer](https://github.com/firecracker-microvm/firecracker-containerd/blob/e80d4cc83f1207c0b9cd5eb1e5bacc67e55e6415/runtime/runc_jailer.go#L195-L222). This fix is extremely delicate, but I am okay with it for now as there are lots of other changes being made to the jailer at the moment and I don't want to create unnecessary conflicts. If you update with `filepath.Base(drive.stubPath)` and it works, I'll just go ahead and create a followup issue for finding a better solution.

Reply:
Great, thanks for explanation.